agent/internal/client/base_client.go

```diff
@@ -123,3 +146,1042 @@ func (c *baseCommunicator) createCedarGRPCConn(ctx context.Context, comm Communi
+	ctx, cancel = context.WithDeadline(ctx, now.Add(10*time.Minute))
+	defer cancel()
+	backupTimer := time.NewTimer(15 * time.Minute)
+	defer backupTimer.Stop()
+	doneChan := make(chan struct{})
+	defer func() {
+		close(doneChan)
+	}()
+	go func() {
+		defer recovery.LogStackTraceAndExit("backup timer")
+		select {
+		case <-ctx.Done():
+			grip.Info("request completed or task ending, stopping backup timer thread")
+			return
+		case t := <-backupTimer.C:
+			grip.Alert(message.Fields{
+				"message":  "retryRequest exceeded 15 minutes",
+				"start":    now.String(),
+				"end":      t.String(),
+				"task":     taskData.ID,
+				"messages": msgs,
+			})
+			cancel()
+			return
+		case <-doneChan:
+			return
+		}
+	}()
+	resp, err := c.retryRequest(ctx, info, &payload)
+	if err != nil {
+		return utility.RespErrorf(resp, "problem sending %d log messages for task %s: %s", len(msgs), taskData.ID, err.Error())
+	}
+	defer resp.Body.Close()
+	return nil
+}
+
+// SendTaskResults posts a task's results, used by the attach results operations.
+func (c *baseCommunicator) SendTaskResults(ctx context.Context, taskData TaskData, r *task.LocalTestResults) error {
+	if r == nil || len(r.Results) == 0 {
+		return nil
+	}
+
+	info := requestInfo{
+		method:   http.MethodPost,
+		taskData: &taskData,
+		version:  apiVersion1,
+	}
+	info.setTaskPathSuffix("results")
+	resp, err := c.retryRequest(ctx, info, r)
+	if err != nil {
+		return utility.RespErrorf(resp, "problem adding %d results to task %s: %s", len(r.Results), taskData.ID, err.Error())
+	}
+	defer resp.Body.Close()
+	return nil
+}
+
+// GetPatch tries to get the patch data from the server in json format,
+// and unmarhals it into a patch struct. The GET request is attempted
+// multiple times upon failure. If patchId is not specified, the task's
+// patch is returned
+func (c *baseCommunicator) GetTaskPatch(ctx context.Context, taskData TaskData, patchId string) (*patchmodel.Patch, error) {
+	patch := patchmodel.Patch{}
+	info := requestInfo{
+		method:   http.MethodGet,
+		taskData: &taskData,
+		version:  apiVersion1,
+	}
+	suffix := "git/patch"
+	if patchId != "" {
+		suffix = fmt.Sprintf("%s?patch=%s", suffix, patchId)
+	}
+	info.setTaskPathSuffix(suffix)
+	resp, err := c.retryRequest(ctx, info, nil)
+	if err != nil {
+		return nil, utility.RespErrorf(resp, "failed to get patch for task %s: %s", taskData.ID, err.Error())
+	}
+	defer resp.Body.Close()
+
+	if err = utility.ReadJSON(resp.Body, &patch); err != nil {
+		return nil, errors.Wrapf(err, "problem parsing patch response for %s", taskData.ID)
+	}
+
+	return &patch, nil
+}
+
+// GetCedarConfig returns the cedar service information including the base URL,
+// URL, RPC port, and credentials.
+func (c *baseCommunicator) GetCedarConfig(ctx context.Context) (*apimodels.CedarConfig, error) {
+	cc := &apimodels.CedarConfig{}
+
+	info := requestInfo{
+		method:  http.MethodGet,
+		version: apiVersion1,
+		path:    "agent/cedar_config",
```

Comment:
This method can't be consolidated because [the REST v1 route](https://github.com/evergreen-ci/evergreen/blob/9acf7956188f12cbd024447c69694d9a390eb02d/service/api.go#L788) does not have pod authentication. It seems like `GetCedarConfig` should be split between the two communicators until the REST v2 host route (`/hosts/{host_id}/agent/cedar_config`) is implemented.

Reply:
I've migrated the route to v2 now using middleware that is applicable for both pods and hosts.  I was trying to consolidate this route because `GetCedarConfig` is only called internally in the base client in private functions which don't have access to host/pod specific implementations.